core/arch/arm/kernel/ftrace.c

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+#include <assert.h>
+#include <kernel/ftrace.h>
+#include <kernel/thread.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <printk.h>
+#include <string.h>
+#include <trace.h>
+
+#define MIN_FTRACE_BUF_SIZE	512
+#define MAX_HEADER_STRLEN	128
+
+void ta_fbuf_init(struct ta_head *ta_head, struct tee_ta_session *s)
+{
+	struct ftrace_buf *fbuf = NULL;
+	size_t fbuf_size = 0;
+	int count = 0;
+
+	if (ta_head->ftrace_buf_end > ta_head->ftrace_buf_start)
+		fbuf_size = ta_head->ftrace_buf_end - ta_head->ftrace_buf_start;
+
+	if (fbuf_size >= MIN_FTRACE_BUF_SIZE) {
+		fbuf = (struct ftrace_buf *)ta_head->ftrace_buf_start;
+		fbuf->head_off = sizeof(struct ftrace_buf);
+		count = snprintk((char *)fbuf + fbuf->head_off,
+				 MAX_HEADER_STRLEN,
+				 "Function graph for TA: %pUl @ %p\n",
+				 (void *)&s->ctx->uuid, (void *)ta_head);
+		assert(count < MAX_HEADER_STRLEN);
+
+		fbuf->buf_off = fbuf->head_off + count;
+		fbuf->curr_size = 0;
+		fbuf->max_size = fbuf_size - sizeof(struct ftrace_buf) - count;
+
+		s->fbuf_paddr = virt_to_phys(fbuf);
+		s->fbuf_size = fbuf_size;
+	} else {
+		DMSG("ftrace buffer too small");
+
+		s->fbuf_paddr = 0;
+		s->fbuf_size = 0;
+	}
+}
+
+void ta_fbuf_dump(struct tee_ta_session *s)
+{
+	struct ftrace_buf *fbuf = NULL;
+	struct mobj *mobj = NULL;
+	char *va = NULL;
+	uint32_t ret = 0;
+	uint32_t dump_size = 0;
+	struct thread_param params[3] = { };
+
+	if (!s->fbuf_size)
+		return;
+
+	if (!core_mmu_add_mapping(MEM_AREA_TA_RAM, s->fbuf_paddr,
+				  s->fbuf_size)) {
+		EMSG("Failed to map TA ftrace buffer, pa: %lx, size: %lx",
+			s->fbuf_paddr, s->fbuf_size);
+		return;
+	}
+
+	fbuf = phys_to_virt(s->fbuf_paddr, MEM_AREA_TA_RAM);
+	if (!fbuf || !fbuf->curr_size || (fbuf->curr_size > fbuf->max_size)) {
+		EMSG("Ftrace initial buffer check failed");
```

Comment:
It seems we reach here when the feature is enabled but no data has been collected (such as, no file compiled with `-pg`). I think this corresponds to (!fbuf->curr_size).

If my analysis is correct, then I'd rather have a `DMSG("Ftrace buffer empty (no TA file compiled with -pg?")` right after the (!fbuf->fbuf_size) test at the beginning of the function. That would be more helpful to the user.

Reply:
Yes your analysis is correct. I will rather remove `!fbuf->curr_size` from here, so that we have a file in `/tmp` folder with header only which implicitly means that no file has been compiled with `-pg`. Also, will add separate `DMSG` for `!fbuf->curr_size` as you have suggested.